include/kernel.h

```diff
@@ -5178,6 +5178,516 @@ extern void k_mem_domain_remove_thread(k_tid_t thread);
+	 *
+	 * When provided this indicates that the stop transition
+	 * function may suspend.
+	 */
+	K_ONOFF_SERVICE_SUSPENDABLE_STOP = BIT(3),
+
+	/**
+	 * @brief Flag passed to k_onoff_service_init().
+	 *
+	 * When provided this indicates that the reset operation is
+	 * safe to invoke from ISR context.
+	 */
+	K_ONOFF_SERVICE_ISR_CALLABLE_RESET = BIT(4),
+
+	/**
+	 * @brief Flag passed to k_onoff_service_init().
+	 *
+	 * When provided this indicates that the reset operation may
+	 * suspend.
+	 */
+	K_ONOFF_SERVICE_SUSPENDABLE_RESET = BIT(5),
+
+	/* Internal use. */
+	K_ONOFF_SERVICE_HAS_ERROR = BIT(6),
+
+	/* This and higher bits reserved for internal use. */
+	K_ONOFF_SERVICE_INTERNAL_BASE = BIT(7),
+};
+
+/* Forward declaration */
+struct k_onoff_service;
+
+/**
+ * @brief Signature used to notify an on-off service that a
+ * transition has completed.
+ *
+ * The functions provided for this type are non-suspendable and
+ * isr-callable.
+ *
+ * @param srv the service for which transition was requested.
+ *
+ * @param res the result of the transition.  This shall be zero on
+ * success, or a negative error code.  If an error is indicate the
+ * service shall enter an error state.
+ */
+typedef void (*k_onoff_service_notify_fn)(struct k_onoff_service *srv,
+					  int res);
+
+/**
+ * @brief Signature used by service implementations to effect a
+ * transition.
+ *
+ * Service definitions use two function pointers of this type to be
+ * notified that a transition is required, and a third optional one to
+ * reset service state.
+ *
+ * The start function will only be called from the off state.
+ *
+ * The stop function will only be called from the on state.
+ *
+ * The reset function may only be called when
+ * k_onoff_service_has_error() returns true.
+ *
+ * Functions will be invoked from interrupt context only if the
+ * corresponding isr-capable flag is specified in the service
+ * configuration.
+ *
+ * @param srv the service for which transition was requested.
+ *
+ * @param notify the function to be invoked when the transition has
+ * completed.  The callee shall capture this parameter to notify on
+ * completion of asynchronous transitions.  If the transition is not
+ * asynchronous, notify shall be invoked before the transition
+ * function returns.
+ */
+typedef void (*k_onoff_service_transition_fn)(struct k_onoff_service *srv,
+					      k_onoff_service_notify_fn notify);
+
+/**
+ * @brief State associated with an on-off service.
+ *
+ * No fields in this structure are intended for use by service
+ * providers or clients.  The state is to be initialized once, using
+ * k_onoff_service_init(), when the service provider is initialized.
+ * In case of error it may be reset through the
+ * k_onoff_service_reset() API.
+ */
+struct k_onoff_service {
+	/* List of clients waiting for completion of reset or
+	 * transition to on.
+	 */
+	sys_slist_t clients;
+
+	/* Function to invoke to transition the service to on. */
```

Comment:
Any particular reason not to put these in an API structure and use the device model?

Reply:
This doesn't provide a device abstraction.  It's an implementation of a data structure where a small part of the behavior is specific to a particular service, specified by injecting functions though the service configuration data.

In many cases user code won't be looking to access the service directly.  E.g. a clock driver might have an onoff service object as part of the driver state, and would delegate its public on/off API methods to calls to the request and release operations given that service object.  There wouldn't even be any devicetree elements associated with the service.

In other cases there would be devicetree information used to generate the service object.  I expect a power rail to be described in devicetree as something like:
```
   sensor-rail: sensor-rail {
      compatible = "regulator,gpio";
      rail-gpios = <&gpio1 8 GPIO_ACTIVE_LOW>;
      on-delay = <500>; /* ns */
   };

   sht3xd@3c {
      /* ... */
      power-rail = &sensor-rail;
   };
```
Ideally this would produce a service object definition that could be referenced from the sensor driver's device power management functions, which would be populated with references to generic gpio_regulator_onoff_{start,stop} functions.  Even there there's no real motivation to make the rail a device; we just need some way for the sensor driver to get a reference to the service object.

For that to be done automatically there will have to be devicetree scripting enhancements.  It may well turn out that to be practical in the near term there'd have to be a device associated with the `gpio,regulator` compatible that would implement a generic "regulator" API.

I hope to prototype something like this within the topic-gpio branch in the next few days.

Reply:
Thanks for the response! It seemed like it was probably somewhat orthogonal, but it's good to have a clear answer like this for the record.